### DISCUSSION TITLE
Using SymmetricEncryptionClientSide

### DISCUSSION THREAD

**Discussion by kakedis** [sqlalchemy/sqlalchemy#8268] [Usage Questions]
Hi, I'm trying to follow example 1 in the Symmertic Encryption Client Side wiki; https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionClientSide, but I'm having issues using the hybrid_property (decrypt) in a search query. When I do the filter by the value can't be match. I'm able to encrypt the value when inserting (adding) the model to my PostgreSQL db and when I print the hybrid_property it works fine and the value gets decrypted. However, it doesn't want to work in a query. My query looks something like this;
''' hp is the hybrid_property value '''
`u2 = User.query.filter(User.hp == 'somevalue').first()`
I've also tried it within a session;
`u2 session.query(User).filter_by(User.hp=='somevalue').first()`

However, I'm unable to find a match. Is the example in the wiki still relevant? 

Thanks!

**Comment by zzzeek** [comment]
hi there -

works fine on this end, just had to update the hexlify call.   

Here's complete example

```py
import binascii
import uuid

from Crypto.Cipher import AES

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import Comparator
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Session

key = uuid.uuid4().bytes
"""The encryption key.   Random for this example."""


nonce = uuid.uuid4().bytes
"""for WHERE criteria to work, we need the encrypted value to be the same
each time, so use a fixed nonce if we need that feature.
"""

def aes_encrypt(data):
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    data = data + (" " * (16 - (len(data) % 16)))
    return cipher.encrypt(data.encode("utf-8")).hex()


def aes_decrypt(data):
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    return cipher.decrypt(binascii.unhexlify(data)).decode("utf-8").rstrip()

Base = declarative_base()

class User(Base):
    __tablename__ = 'user'

    id = Column(Integer, primary_key=True)
    encrypted_value = Column(String, nullable=False)

    @hybrid_property
    def value(self):
        return aes_decrypt(self.encrypted_value)

    @value.setter
    def value(self, value):
        self.encrypted_value = aes_encrypt(value)

    class encrypt_comparator(Comparator):
        def operate(self, op, other, **kw):
            return op(
                self.__clause_element__(), aes_encrypt(other),
                **kw
            )

    @value.comparator
    def value(cls):
        return cls.encrypt_comparator(
                    cls.encrypted_value
                )

e = create_engine('postgresql://scott:tiger@localhost/test', echo='debug')

Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

# attribute set
u1 = User(value="some value")
s.add(u1)
s.commit()

# comparison
u2 = s.query(User).filter_by(value="some value").first()
assert u1 is u2

# attribute get
assert u1.value == "some value"
```

**Reply by zzzeek** [reply]
specifically here's the query at the end:

```sql
2022-07-18 14:34:30,548 INFO sqlalchemy.engine.Engine SELECT "user".id AS user_id, "user".encrypted_value AS user_encrypted_value 
FROM "user" 
WHERE "user".encrypted_value = %(encrypted_value_1)s 
 LIMIT %(param_1)s
2022-07-18 14:34:30,548 INFO sqlalchemy.engine.Engine [generated in 0.00009s] {'encrypted_value_1': '97ff5fc038141919c0adf50b7f433239', 'param_1': 1}
2022-07-18 14:34:30,549 DEBUG sqlalchemy.engine.Engine Col ('user_id', 'user_encrypted_value')
2022-07-18 14:34:30,549 DEBUG sqlalchemy.engine.Engine Row (1, '97ff5fc038141919c0adf50b7f433239')

```

you want to see what your program does there with echo='debug'

**Reply by kakedis** [reply]
Thanks for the quick response, I really appreciate it! 

After debugging I'm able to see some additional information. This is the debug output: 
``WHERE c.ph_1 = %(ph_1_1)s 
LIMIT %(param_16)s
2022-07-18 13:39:42,450 INFO sqlalchemy.engine.base.Engine {'ph_1_1': 'ytOZZ6eDGnT4v8CwjZE4+p2euKGuwigiqBNvGFKz5yU='}
{'ph_1_1': 'ytOZZ6eDGnT4v8CwjZE4+p2euKGuwigiqBNvGFKz5yU='}
2022-07-18 13:39:42,453 DEBUG sqlalchemy.engine.base.Engine Col ('c_ph_1')
Col ('c_ph_1')``

The above query results in `None`

From what I can see it seems like the value(parameter) I'm trying to match with get encrypted when running the query. I expected ph_1_1 to be an unencrypted string, because my hybrid_property is decrypted. Am I suppose to be matching on an encrypted values or decrypted values in my query? We way I encrypt will mean that the same value will look different with each encryption, which means I won't be able to match on encrypted value. Is there a way to use decrypted values in the query? Here's my code. I use the python pycrypto package to handle encryption;

```# DB Model
class Contact(ResourceMixin, db.Model):
    __tablename__ = 'contacts'`
    ph_1 = db.Column(db.String(20), nullable=False, server_default='')

    @hybrid_property
    def pn(self):
        crypto_key = app.config['CRYPTO_SECRET_KEY']
        cipher = AESCipher(crypto_key)
        return cipher.decrypt(self.ph_1)

    @pn.setter
    def pn(self, pn):
        crypto_key = app.config['CRYPTO_SECRET_KEY']
        cipher = AESCipher(crypto_key)
        self.ph_1 = cipher.encrypt(pn)

    class encrypt_comparator(Comparator):
        def operate(self, op, other, **kw):
            crypto_key = app.config['CRYPTO_SECRET_KEY']
            cipher = AESCipher(crypto_key)
            return op(
                self.__clause_element__(), cipher.encrypt(other),
                **kw
            )

    @pn.comparator
    def pn(cls):
        return cls.encrypt_comparator(
            cls.ph_1)
            
# Here's the code usage;

# Set new value
new_ph = 12345

# Add record to db table 
contact = Contact(
    pn=new_ph,
)
db.session.add(contact)
db.session.flush()
db.session.commit()

print(new_ph) # restult: 12345

print((contact.ph_1) # result: Md4ddcJg4pTlR5rXxh3ipKZSR4avVT9BwkY9PEtmfxkvc=

print(contact.pn) # result: 12345

# Perform query 
# I'm expecting the pn value to be a decrypted value for ph_1 and match with an decrypted value for new_ph. If encrypted they won't match.

new_contact = db.session.query(Contact).filter_by(pn=new_ph).first()
print(new_contact) # result: None

# Another few unsuccessful queries. I expect Contact.pn to be an decrypted value as well new_ph. If encrypted they won't match.

alt_new_contact = Contact.query.filter(Contact.pn==new_ph).first()
print(alt_new_contact) # result: None
````
It seems like I'm trying to match different encrypted values. I was hoping to match on decrypted values as the example above. How would I achieve this?

Thanks in advance!

**Reply by zzzeek** [reply]
if your encryption and decryption occurs on the client only, and not in a SQL statement, and the data that's in a target column in the database is encrypted, then a WHERE expression ultimately has to compare the encrypted values, because the database is going to match a string you pass into the statement to the characters that are directly in the column.     This also means the kind of encryption you're using has to have a deterministic string representation each time for a given value.  so if there's some kind of "salt" or something that is changing each time, that would have to be the same each time for a given literal value so that you get the identical encrypted string each time.

the debug output looks like it's the correct idea (if not the correct specifics), in that the parameter passed in for comparison is encrypted.

the code you have looks just like the example so try running my example directly and see if that works.

**Reply by kakedis** [reply]
Apologies for only getting back to you now, but I've been trying to find ways to implement the client side encryption, but running into a bunch of issues. I've decided to go with the 2nd example using TypeDecorator, because it seems like a simpler and cleaner solution. However, I'm not getting the desired results, due to either `str` or `bytes` issues. I'm hoping you will be quick to point out where I'm going wrong. This is what I have;

```
nonce = uuid.uuid4().bytes


class SymmetricEncryptionClientSide:

    def sym_encrypt(self, data):
        key = app.config['KEY']  # example str: 'we5RdfHfz2V1/2eL'
        cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
        data = data + (" " * (16 - (len(data) % 16)))
        return binascii.hexlify(cipher.encrypt(data.encode("utf-8")))

    def sym_decrypt(self, data):
        key = app.config['KEY']  # example str: 'we5RdfHfz2V1/2eL'
        cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
        return cipher.decrypt(binascii.unhexlify(data)).decode("utf-8").rstrip()


class EncryptedValue(TypeDecorator):
    impl = String

    def process_bind_param(self, value, dialect):
        cipher = SymmetricEncryptionClientSide()
        return cipher.sym_encrypt(value)

    def process_result_value(self, value, dialect):
        cipher = SymmetricEncryptionClientSide()
       # I perform a try catch because I have existing data in my database that is not encrypted 
        try:
            return cipher.sym_decrypt(value)
        except:
            return value

# DB Model and field
class Contact(ResourceMixin, db.Model):
    __tablename__ = 'contacts'
   id = db.Column(db.Integer, primary_key=True)
   email = db.Column(EncryptedValue(), nullable=False, server_default='')
```
The only real difference I think between your example and my code is that you have use uuid value turned into bytes for a key. I have a string for a key and if I try to add `.bytes` I get a `'str' object has no attribute 'bytes'` error. The other difference is for the database field you give it a column name; `db.Column("encrypted_value", EncryptedValue(),`. I omitted this, because I got query errors say that the field did not exist so I removed it. 

When I use the above code and enter a value into the email field into the database I get the following Type error;
`TypeError("Object type %s cannot be passed to C code" % type(data))`

To overcome this error I added `.encode("utf-8")` to my key;  `key = app.config['KEY'] ` and it seemed to have worked and there is a encrypted value in the database that looks like;
`\x64326164356666383433623638363262656530613164636430336461356132623663616538363463306337326638643130643466396236656530353234643636`
when I make another entry using the same value I can confirm that the string is the same and it saves to the db as; `\x64326164356666383433623638363262656530613164636430336461356132623663616538363463306337326638643130643466396236656530353234643636`

So far it seems to be working. However, I'm having trouble on the decrypting side, which makes me think the encrypting is also not working correctly? When I look up the same record by id using a query like the following;
`contact_test = Contact.query.filter(Contact.id == record_id).first()` and return the email value it's returned as; 
`\x64326164356666383433623638363262656530613164636430336461356132623663616538363463306337326638643130643466396236656530353234643636` which I didn't expect. I was expecting the email value to be decrypted when returned from the db in the query not to return the encrypted string. 
I then perform another query looking for the email value;
`contact_test = Contact.query.filter(and_(Contact.id == 28923, Contact.email == 'johnfoe@hotmail.com')).first()` but I get the following error;
`psycopg2.errors.UndefinedFunction: operator does not exist: character varying = bytea
LINE 12: WHERE contacts.id = 28923 AND contacts.email = '\x3639346336...`

Are you able to assist me and tell me where I'm going wrong? Basically it seems to encrypting correctly, with the same value being the same encrypted string, but I can't seem to decrypt the value when returning it from the db. 

Thanks for the help!

**Reply by zzzeek** [reply]
so I'm going away for a week, and there's a lot going on up there, so I would suggest if you started with the example from the wiki, first make sure that example works completely.  then just iteratively change one thing at a time, to get to the program you actually have, so that you can isolate exactly when it stops working.     maybe you already did that, so I will try to take a look in the coming days if not next week.

**Reply by kakedis** [reply]
Thanks @zzzeek, I really appreciate your time helping me with this. I did start with the example on the wiki, but eventually I have to integrate it into my existing code and use existing tables/data. When using some of the example code I ran into errors and had to tweak it to make it work in my project. I also tried and isolated the issue and it seems to be an issue with my decrypting, but I'm worried the issue is actually before that. When I perform the encrypt an encrypted value string as shown below is saved to the db. It's always the same string if the data value is the same, which is exactly what I want. So this seems to work fine. 
`\x6230356534373164303264393734386566663733306637313835613131363062`

The problem comes in with the decrypting. When I try and decrypt this value I get this error message:

`binascii.Error: Non-hexadecimal digit found` 

which happens somewhere in this line of code; 
`cipher.decrypt(binascii.unhexlify(data)).decode("utf-8").rstrip()`

As mentioned I'm using the same code to encrypt and decrypt as the example and my code seems to be encrypting correctly, but I'm not sure, because the decrypt is not working.

The only two differences between the example code and the code I have is that my key is a string; `key.encode("utf-8")` instead of `uuid.uuid4().bytes`, which doesn't seem to be the problem, because I get the same result. The other difference is my field column doesn't have a column name. It's;
`email = db.Column(EncryptedValue(), nullable=False, server_default='')` without the column name. I got some errors when trying to give the existing column a name. I'm trying to implement this on an existing column that already has data. Besides these two differences everything is the same. Here's what I'm doing with encrypting and decrypting;

```
nonce = uuid.uuid4().bytes

def sym_encrypt(data):
    key = app.config['KEY'].encode("utf-8")
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    data = data + (" " * (16 - (len(data) % 16)))
    return binascii.hexlify(cipher.encrypt(data.encode("utf-8")))


def sym_decrypt(data):
    key = app.config['KEY'].encode("utf-8")
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    return cipher.decrypt(binascii.unhexlify(data)).decode("utf-8").rstrip()


class EncryptedValue(TypeDecorator):
    impl = String

    def process_bind_param(self, value, dialect):
        return sym_encrypt(value)

    def process_result_value(self, value, dialect):
        try:
            return sym_decrypt(value)
        except:
            return value
```
I'm a little lost and not sure if it's actually an issue with the decrypt or if its the encrypt that is causing the issue. I hoping you might see something obvious I'm doing wrong.

Thanks again!

**Reply by CaselIT** [reply]
The issue here seems not releted to sqlalchemy, but on the way you are decrypting things.
You mentioned having non encrypted data in the db, so that may be causing the issue.

Without a runnable example it's hard to suggest anything more

**Reply by kakedis** [reply]
@CaselIT The issue was not with the decryption. It was with the database field using the TypeDecorator. I had it like this;

`email_field = db.Column(EncryptedValue(), nullable=False, server_default='')`

However, after chaging it to;

`email_field = db.Column(EncryptedValue(40), nullable=False, server_default='')`

I was omitting the 40, which I assume is the length. I'm not sure if this was the actual issue, but it seems to be working now. 

@zzzeek Does this make sense that this could've caused the issue? Thanks for the help on this I really appreciate the responses you provided.

**Reply by zzzeek** [reply]
for Python 3, your encrypt should look like this:

```py
def aes_encrypt(data):
    cipher = AES.new(key, AES.MODE_EAX, nonce=nonce)
    data = data + (" " * (16 - (len(data) % 16)))
    return cipher.encrypt(data.encode("utf-8")).hex()

```

given that, a length of 40 is too short.   using unspecified length means a VARCHAR of arbitrary size and I can't think of a reason this wouldn't work other than the comparison failing if you are using unhexlify which produces bytes, which will fail to compare to VARCHAR correctly.

**Reply by kakedis** [reply]
Thank you @zzzeek, I got it to work and the decryption and encryption is working as intended. However, I do have 2 follow up questions using client side encryption and SQLAlchemy. If these questions are not appropriate for this thread please let me know.

1. I'm using the `TypeDecorator` method for client encryption. Is it possible to use these fields in hybrid_properties and expressions? I'm trying to combine two encrypted fields (using the `TypeDecorator` for encryption); firstname and lastname to return a single value; full name, e.g. 
```
@hybrid_property
def fullname(self):
return self.firstname + ' ' + self.lastname
```
  However, I get a AttributeError when trying to do this. Is there a way to create a new field property combining encrypted fields?

2. I'm able to use the `TypeDecorator` encrypted fields in SQLAlchemy filter queries, which is great. However, I have some more complex use cases, which I'm struggling with. I'm trying to use the encrypted fields in ilike queries, but not having any luck. I'm trying to perform some searching on the server side using Datatables. Here's an example query;
```
pattern = '%{}%'.format(search)
        filtered = total.filter(or_(
            Contact.firstname.ilike(pattern),
            Contact.lastname.ilike(pattern),
        ))
 ```
 In the query above both the firstname and lastname fields are `TypeDecorator` encrypted. As-is the above query does not work. If the fields are not encrypted it works fine. If I change the query to;
 ```
 pattern = '%{}%'.format(search)
        filtered = total.filter(or_(
            Contact.firstname == search,
            Contact.lastname == search,
        ))
```
I'm able to return one result if its exactly the same value being searched. There's two issues with this;

a) I would like to search for the pattern ideally instead of just the exact match. This is not a deal breaker, but better for usability. It would be great if I can use ilike or like operators. 
b) The query always only return 1 result even if there's multiple results with the same firstname or lastname being search. 

Do you have any ideas on how I can handle the above query to address a) and b) or at least b)? 

Thanks in advance!

**Reply by zzzeek** [reply]
> Thank you @zzzeek, I got it to work and the decryption and encryption is working as intended. However, I do have 2 follow up questions using client side encryption and SQLAlchemy. If these questions are not appropriate for this thread please let me know.
> 
>     1. I'm using the `TypeDecorator` method for client encryption. Is it possible to use these fields in hybrid_properties and expressions? I'm trying to combine two encrypted fields (using the `TypeDecorator` for encryption); firstname and lastname to return a single value; full name, e.g.
> 
> 
> ```
> @hybrid_property
> def fullname(self):
> return self.firstname + ' ' + self.lastname
> ```
> 
> However, I get a AttributeError when trying to do this. Is there a way to create a new field property combining encrypted fields?

just what you're doing should work fine, would need more details

> 
>     2. I'm able to use the `TypeDecorator` encrypted fields in SQLAlchemy filter queries, which is great. However, I have some more complex use cases, which I'm struggling with. I'm trying to use the encrypted fields in ilike queries, but not having any luck. I'm trying to perform some searching on the server side using Datatables. Here's an example query;
> 
> 
> ```
> pattern = '%{}%'.format(search)
>         filtered = total.filter(or_(
>             Contact.firstname.ilike(pattern),
>             Contact.lastname.ilike(pattern),
>         ))
> ```
> 
> In the query above both the firstname and lastname fields are `TypeDecorator` encrypted. As-is the above query does not work. If the fields are not encrypted it works fine. If I change the query to;

that's not going to be possible, because those fields are encrypted on the server and you aren't using server side decryption.  you would need to be using some fancy kind of homomorphic encryption which I dont think is a thing for this kind of operation.

> 
> ```
> pattern = '%{}%'.format(search)
>        filtered = total.filter(or_(
>            Contact.firstname == search,
>            Contact.lastname == search,
>        ))
> ```
> 
> I'm able to return one result if its exactly the same value being searched. There's two issues with this;
> 
> a) I would like to search for the pattern ideally instead of just the exact match. This is not a deal breaker, but better for usability. It would be great if I can use ilike or like operators. b) The query always only return 1 result even if there's multiple results with the same firstname or lastname being search.
> 
> Do you have any ideas on how I can handle the above query to address a) and b) or at least b)?

things would be easier if you were using server side encryption. there's no way to search encrypted fields without making some kind of index separately, and that index would have the unencrypted data (or if you built a fancy homomorphic encryption scheme, which is above my pay grade).

**Reply by kakedis** [reply]
Thanks @zzzeek 
There's some more info on the hybrid_property;
```
@hybrid_property
    def dec_name(self):
        return self.firstname + " " + self.lastname

@dec_name.expression
    def dec_name(cls):
        return cls.firstname + " " + cls.lastname
```
this return a result as follows;
`ac3c4f6bbb4d9b5740425cced0f684f7c6730725bb4d9b5740425cced0f684f7a23c4225bb4d9b5740425cced0f684f7`, which seems to 3 parts; 
the firstname encrypted value; `ac3c4f6bbb4d9b5740425cced0f684f7`, an encryption, of what I assume is the space between first and last name; `c6730725bb4d9b5740425cced0f684f7` and the lastname encrypted value; `a23c4225bb4d9b5740425cced0f684f7`. When I only return the first name or the last name in the hybrid property and expression the correct decrypted value gets returned, e.g;
```
 @hybrid_property
    def dec_name(self):
        return self.firstname

@dec_name.expression
    def dec_name(cls):
        return cls.firstname
```
The value comes back correct an decrypted as `John`. So I'm bit confused why the first example doesn't work and returns a string with the encrypted values and have the space encrypted. Any thoughts on this?

As for actually encryption, you mention that things will be easier using server side encryption. Do you suggest I look at the <a href="https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionServerSide">SymmetricEncryptionServerSide</a> wiki and examples you created? I just want to make sure you are referring to those methods when you are speaking of server side encryption.  I noticed you say its MySQL specific, but I'm using PostgreSQL. 

Thanks again!

**Reply by zzzeek** [reply]
the SQL experssion aspect of the hybrid property will not work, for the same reason, these values are encrypted.  you can't concatenate them together in a SQL expression.  you'd have to load Contact objects first into memory, then look at some_contact.dec_name, and you'll get the value because .firstname / .lastname will have been unencrypted into memory.

so yes you can look into the server side encryption example also if you really need to do SQL like/concat/etc on the server side.  you can look into using [pgcrypto](https://www.postgresql.org/docs/current/pgcrypto.html) for that.
